flocker/ca/_validation.py

```diff
@@ -0,0 +1,108 @@
+# Copyright ClusterHQ Inc.  See LICENSE file for details.
+
+"""
+TLS context factories used to validate the various kinds of
+certificates generated by the CA.
+"""
+
+from OpenSSL.SSL import VERIFY_PEER, VERIFY_FAIL_IF_NO_PEER_CERT
+
+from zope.interface import implementer
+
+from twisted.web.iweb import IPolicyForHTTPS
+from twisted.internet.ssl import (
+    optionsForClientTLS, CertificateOptions,
+    )
+
+from pyrsistent import PRecord, field
+
+
+@implementer(IPolicyForHTTPS)
+class ControlServicePolicy(PRecord):
+    """
+    HTTPS TLS policy for validating the control service's identity, and
+    providing the HTTP client's identity.
+
+    :ivar Certificate ca_certificate: The certificate authority's
+        certificate.
+
+    :ivar PrivateCertificate client_certificate: Client's
+        certificate/private key pair.
+    """
+    ca_certificate = field(mandatory=True)
+    client_certificate = field(mandatory=True)
+
+    def creatorForNetloc(self, hostname, port):
+        return optionsForClientTLS(u"control-service",
+                                   trustRoot=self.ca_certificate,
+                                   clientCertificate=self.client_certificate)
+
+
+class _ClientContextFactory(object):
+    """
+    Context factory that validates various kinds of clients that can
+    connect to the control service.
+    """
+    def __init__(self, ca_certificate, control_credential, prefix):
+        """
+        :param Certificate ca_certificate: The certificate authority's
+            certificate.
+
+        :param ControlCredential control_credential: The control service's
+            credentials.
+
+        :param bytes prefix: The required prefix on certificate common names.
+        """
+        key = control_credential.credential.keypair.keypair.original
+        certificate = control_credential.credential.certificate.original
```

Comment:
Blarghy blargy blarg blarh blaaaa...

How about giving `ControlCredential` a method which returns a suitably constructed `CertificateOptions` and this code can call it instead of reaching way down into `ControlCredential`'s bowels for these pieces?

Trying to resolve the meaning of these two lines, I also noticed that `FlockerCredential.keypair` is documented as a `FlockerKeyPair` - which is not actually defined anywhere.